src/job-exporter/src/collector.py

```diff
@@ -267,13 +294,76 @@ class GpuCollector(Collector):
     cmd_histogram = Histogram("cmd_nvidia_smi_latency_seconds",
             "Command call latency for nvidia-smi (seconds)")
 
-    cmd_timeout = 3 # 99th latency is 0.97s
+    cmd_timeout = 60 # 99th latency is 0.97s
 
-    def __init__(self, name, sleep_time, atomic_ref, iteration_counter, gpu_info_ref):
+    def __init__(self, name, sleep_time, atomic_ref, iteration_counter, gpu_info_ref, zombie_info_ref):
         Collector.__init__(self, name, sleep_time, atomic_ref, iteration_counter)
         self.gpu_info_ref = gpu_info_ref
+        self.zombie_info_ref = zombie_info_ref
+
+    @staticmethod
+    def get_container_id(pid):
+        """ return two values, the first one is if we found the corresponding
+        container_id, the second one is the container_id if found """
+        path = "/proc/%d/cgroup" % (pid)
+        if not os.path.isfile(path):
+            return False, ""
+
+        with open(path) as f:
+            content = f.read()
+
+        for line in content.split("\n"):
+            line = line.strip()
+            if "pids" in line and "/docker/" in line:
+                parts = line.split("/docker/")
+                if len(parts) == 2 and re.match(u"[0-9a-f]+", parts[1]):
+                    return True, parts[1]
+
+        return False, ""
+
+    @staticmethod
+    def convert_to_metrics(gpu_info, zombie_info, pid_to_cid_fn):
+        """ This fn used to convert gpu_info & zombie_info into metrics, used to make
+        it easier to do unit test """
+        core_utils = gen_gpu_util_gauge()
+        mem_utils = gen_gpu_mem_util_gauge()
+        ecc_errors = gen_gpu_ecc_counter()
+        mem_leak = gen_gpu_memory_leak_counter()
+        external_process = gen_gpu_used_by_external_process_counter()
+        zombie_container = gen_gpu_used_by_zombie_container_counter()
+
+        pids_use_gpu = {} # key is gpu minor, value is an array of pid
+
+        for minor, info in gpu_info.items():
+            core_utils.add_metric([minor], info.gpu_util)
+            mem_utils.add_metric([minor], info.gpu_mem_util)
+            ecc_errors.add_metric([minor, "single"], info.ecc_errors.single)
+            ecc_errors.add_metric([minor, "double"], info.ecc_errors.double)
+            if info.gpu_mem_util > 20 * 1024 * 1024 and len(info.pids) == 0:
```

Comment:
Could 20M be a configurion?

Reply:
This 20M was mentioned by @fanyangCS , I'm not sure if we want it to be configurable or hardcode should be enough.

Reply:
@xudifsd 
In our yarn patch, this is a configuration, so an item with default value seems reasonable.

Reply:
ok